include/engine/datafacade/contiguous_internalmem_datafacade.hpp

```diff
@@ -1070,12 +1070,48 @@ template <> class ContiguousInternalMemoryAlgorithmDataFacade<MLD> : public Algo
         InitializeInternalPointers(allocator->GetLayout(), allocator->GetMemory());
     }
 
-    const partition::MultiLevelPartitionView &GetMultiLevelPartition() const override
+    LevelID GetQueryLevel(NodeID source, NodeID target, NodeID node) const override
     {
-        return mld_partition;
+        return mld_partition.GetQueryLevel(source, target, node);
     }
 
-    const partition::CellStorageView &GetCellStorage() const override { return mld_cell_storage; }
+    LevelID GetHighestDifferentLevel(NodeID first, NodeID second) const override
+    {
+        return mld_partition.GetHighestDifferentLevel(first, second);
+    }
+
+    void ForEachSourceNodes(LevelID level, NodeID node, NodeFnT const &fn) const override
+    {
+        const auto &cell = mld_cell_storage.GetCell(level, mld_partition.GetCell(level, node));
+        auto source = cell.GetSourceNodes().begin();
+        auto durations = cell.GetInDuration(node).begin();
+
+        BOOST_ASSERT(cell.GetSourceNodes().size() == cell.GetInDuration(node).size());
+        BOOST_ASSERT(cell.GetSourceNodes().size() == cell.GetInWeight(node).size());
+
+        for (auto shortcut_weight : cell.GetInWeight(node))
+            fn(*source++, shortcut_weight, *durations++);
+    }
+
+    void ForEachDestinationNodes(LevelID level, NodeID node, NodeFnT const &fn) const override
+    {
+        const auto &cell = mld_cell_storage.GetCell(level, mld_partition.GetCell(level, node));
+        auto destination = cell.GetDestinationNodes().begin();
+        auto durations = cell.GetOutDuration(node).begin();
+
+        BOOST_ASSERT(cell.GetDestinationNodes().size() == cell.GetOutDuration(node).size());
```

Comment:
> sometimes i got 20 nodes and 0 durations and weights

This would indicate that the node was not a source/destination node respectively. Calling `cell.GetOutDuration(node)` on a node that is not actually a source node for the cell will return an empty range. Same for destination nodes and `cell.GetInDuration(node)` if `node` is not a destination node.